### PR TITLE
Issue 229 new ecnt

### LINKED ISSUE #229: Implement aggregation of the new ECNT statistics: SPREAD_MD, MAE, MAE_OERR, BIAS_RATIO, N_GE_OBS, ME_GE_OBS, N_LT_OBS, and ME_LT_OBS
## Describe the Enhancement ##
MET development version 11.0.0-beta4 made several changes to the "stat" output. This issue is to enhance METcalcpy to aggregate these newly added statistics across multiple cases.

- Issue dtcenter/MET#2332 adds 1 new column (SPREAD_MD) to the ECNT line type generated by Ensemble-Stat and the HiRA method in Point-Stat. See [Table 13.2](https://met.readthedocs.io/en/develop/Users_Guide/ensemble-stat.html?highlight=ECNT#id2). The SPREAD_MD is the ensemble mean absolute difference and is an alternative measure of ensemble spread. This column should be aggregated over multiple cases using a weighted average approach, where the weight is defined by the number of matched pairs in the TOTAL column.

- Issue dtcenter/MET#2325 adds 2 new columns (MAE and MAE_OERR) to the ECNT line type generated by Ensemble-Stat and the HiRA method in Point-Stat. See [Table 13.2](https://met.readthedocs.io/en/develop/Users_Guide/ensemble-stat.html?highlight=ECNT#id2). The MAE is the mean absolute error. These columns should be aggregated together over multiple cases using a weighted average approach, where the weight is defined by the number of matched pairs in the TOTAL column.

These MAE's should be added to the list of statistics where 0 is the perfect score.

- Issue dtcenter/MET#2058 adds 5 new columns (BIAS_RATIO, N_GE_OBS, ME_GE_OBS, N_LT_OBS, ME_LT_OBS) to the ECNT line type generated by Ensemble-Stat and the HiRA method in Point-Stat. See [Table 13.2](https://met.readthedocs.io/en/develop/Users_Guide/ensemble-stat.html?highlight=ECNT#id2). All 5 columns are added so that the BIAS_RATIO can be aggregated over multiple runs, as described below.
   - N_GE_OBS is an integer and is aggregated over multiple runs by summing the values.
   - ME_GE_OBS is a float and is aggregated as a weighted average where the weights are defined by the N_GE_OBS column.
   - N_LT_OBS is an integer and is aggregated over multiple runs by summing the values.
   - ME_LT_OBS is a float and is aggregated as a weighted average where the weights are defined by the N_LT_OBS column.
   - BIAS_RATIO is a float >0. It aggregated by first aggregating ME_GE_OBS and ME_LT_OBS and then computing BIAS_RATIO = ME_GE_OBS / absolute value of ME_LT_OBS.

The ME_GE_OBS and ME_LT_OBS columns are mean errors and should added to list of statistics where 0 is the perfect score.
The BIAS_RATIO is the ratio of those 2 ME values, and I suppose a perfect score would be 0/0 which is undefined. This *probably* should go in the list of statistics where 1 is the perfect score... where the amount of high bias equal the amount of low bias. In this way, its similar to the frequency bias (FBIAS) statistic.

Sample data for this issue can be found in the nightly build output in seneca:/d1/projects/MET/MET_regression/develop/NB20221103/MET-develop/test_output (or a more recent date).
And here's a tarfile of that output:
ftp://ftp.rap.ucar.edu/incoming/irap/met_help/NB20221103_develop_test_output.tar.gz

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
Split 2799991 and 2773542

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by TatianaBurek**
## Pull Request Testing ##

- [ ] Describe testing already performed for these changes:
- compared results on the added methods with the statistics calculated by MET

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes or No]**

- [ ] Do these changes include sufficient testing updates? **[Yes or No]**

- [ ] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Development** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: metcalcpy/util/ecnt_statistics.py [hunk 1]
```diff
@@ -263,17 +291,10 @@ def calculate_ecnt_rmse(input_data, columns_names, aggregation=False):
...
-    try:
-        total = get_total_values(input_data, columns_names, aggregation)
-        rmse = math.sqrt(sum_column_data_by_name(input_data, columns_names, 'mse') / total)
-        result = round_half_up(rmse, PRECISION)
-    except (TypeError, ZeroDivisionError, Warning, ValueError):
-        result = None
-    warnings.filterwarnings('ignore')
-    return result
+
+    return weighted_average(input_data, columns_names, 'mse', aggregation)
```
Comment (JohnHalleyGotway) [inline metcalcpy/util/ecnt_statistics.py:297]: @TatianaBurek good idea writing/using a utility function to handle the common computation of weighted averages. I'm a bit suspicious on this one though. I think we need to compute the weighted average of the MSE and the compute the square root of the result and return that. That's what the old code was doing... so you just need to port back over that previous logic.

File: metcalcpy/util/ecnt_statistics.py [hunk 2]
```diff
@@ -339,18 +353,10 @@ def calculate_ecnt_rmse_oerr(input_data, columns_names, aggregation=False):
...
-        total = get_total_values(input_data, columns_names, aggregation)
-        mse_oerr = sum_column_data_by_name(input_data, columns_names, 'mse_oerr') / total
-        rmse_oerr = np.sqrt(mse_oerr)
-        result = round_half_up(rmse_oerr, PRECISION)
-    except (TypeError, ZeroDivisionError, Warning, ValueError):
-        result = None
-    warnings.filterwarnings('ignore')
-    return result
+
+    return weighted_average(input_data, columns_names, 'mse_oerr', aggregation)
```
Comment (JohnHalleyGotway) [inline metcalcpy/util/ecnt_statistics.py:360]: Same comment as above. We need to aggregate the MSE_OERR and the compute the square root of the result.

File: metcalcpy/util/ecnt_statistics.py [hunk 3]
```diff
@@ -404,6 +410,124 @@ def calculate_ecnt_spread_plus_oerr(input_data, columns_names, aggregation=False
...
+            aggregation: if the aggregation on fields was performed
+
+        Returns:
+            calculated ME_GE_OBS as float
+            or None if some data values are missing or invalid
+    """
+    warnings.filterwarnings('error')
+    try:
+        n_ge_obs = sum_column_data_by_name(input_data, columns_names, 'n_ge_obs')
+        me_ge_obs = sum_column_data_by_name(input_data, columns_names, 'me_ge_obs')/n_ge_obs
```
Comment (JohnHalleyGotway) [inline metcalcpy/util/ecnt_statistics.py:477]: I don't think this logic is correct. We need to compute an aggregated ME_GE_OBS as a weighted average where N_GE_OBS defines the weight. That's different from what's being compute here.

File: metcalcpy/util/ecnt_statistics.py [hunk 4]
```diff
@@ -404,6 +410,124 @@ def calculate_ecnt_spread_plus_oerr(input_data, columns_names, aggregation=False
...
+            aggregation: if the aggregation on fields was performed
+
+        Returns:
+            calculated ME_GE_OBS as float
+            or None if some data values are missing or invalid
+    """
+    warnings.filterwarnings('error')
+    try:
+        n_lt_obs = sum_column_data_by_name(input_data, columns_names, 'n_lt_obs')
+        me_lt_obs = sum_column_data_by_name(input_data, columns_names, 'me_lt_obs')/n_lt_obs
```
Comment (JohnHalleyGotway) [inline metcalcpy/util/ecnt_statistics.py:501]: I don't think this logic is correct. We need to compute an aggregated ME_LT_OBS as a weighted average where N_LT_OBS defines the weight. That's different from what's being compute here.

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
@TatianaBurek thanks for working on these updates. I made a handful of comments that require attention. I'll mark this as "Request Changes". Please just re-request my review once you're finished with the next round of updates.

File: metcalcpy/util/ecnt_statistics.py (hunk 3 above)
Comment (TatianaBurek) [reply metcalcpy/util/ecnt_statistics.py:477]: In the common case when total is used as a weight we get the sum of total values, get sum of the column values and divide column sum on total sum (weighted_average method)
Here I did the same but replaced  total with n_ge_obs. Why this is incorrect? How should it be computed?

File: metcalcpy/util/ecnt_statistics.py [hunk 5]
```diff
@@ -134,18 +112,53 @@ def calculate_ecnt_crps_emp_fair(input_data, columns_names, aggregation=False):
...
+    :param column_name: name of the column to be aggregated
+    :param aggregation: if the aggregation on fields was performed
+    :return: aggregated column values or None if some data values are missing or invalid
     """
     warnings.filterwarnings('error')
     try:
         total = get_total_values(input_data, columns_names, aggregation)
-        crps_emp_fair = sum_column_data_by_name(input_data, columns_names, 'crps_emp_fair') / total
-        result = round_half_up(crps_emp_fair, PRECISION)
+        statistic = sum_column_data_by_name(input_data, columns_names, column_name) / total
```
Comment (JohnHalleyGotway) [inline metcalcpy/util/ecnt_statistics.py:154]: @TatianaBurek, yes, good point I suspect there's an issue here as well. It's possible I just don't understand how this code is working. But it looks to me like you're summing up the STATISTIC values and TOTAL counts and dividing the first by the second.

Using R to provide an example of aggregating MAE values, where the weight is defined by the total column:
```
R
MAE = c(5, 10, 8)
TOTAL = c(1000, 1500, 1250)
c("Bad aggregated value using this logic = ", sum(MAE) / sum(TOTAL))
# Prints incorrect value of 0.00613
c("Correct weighted aggregation = ", sum(MAE*TOTAL)/sum(TOTAL))
# Prints correct value of 8
```
Those values are so different, so I'm assuming this IS working, but I'm just not grasping how.

Comment (TatianaBurek) [reply metcalcpy/util/ecnt_statistics.py:154]: the preprocessing of the data  - calculation of the additional statistics , multiplying  by the weight, renaming columns - is happening in agg_stat.py script. For example, here I prepare ecnt data:
https://github.com/dtcenter/METcalcpy/blob/deedf36fc910aee5b970e9fcfb2de199e771bc14/metcalcpy/agg_stat.py#L448
calculate_<statistic> methods use the data that was already multiplied  by the weight

Comment (TatianaBurek) [reply metcalcpy/util/ecnt_statistics.py:154]: The decision on separating preparing and calculating data was made based on the need to improve the speed of bootstraping process. It makes more sense to precalculate base values once and use these values n-replication times during  bootstraping

Comment (JohnHalleyGotway) [reply metcalcpy/util/ecnt_statistics.py:154]: OK, as long as you're confident that the weighted averages are being computed properly, I'll go ahead and approve. Those details just aren't immediately obvious when reviewing the code.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes after Tatiana double-checked and confirmed that the weighted averages are being computed properly.

**Comment by TatianaBurek** [comment]
I checked that me_ge_obs and me_lt_obs are calculated the same way as other  weighted average stats but instead of total as the weight they use n_ge_obs and n_lt_obs.

**Pull request merged**